funfact/algorithm.py

```diff
@@ -101,52 +102,53 @@ class _Factorization(Factorization, ab.AutoGradMixin):
             'Invalid optimization algorithm:\n{e}'
         )
 
-    @ab.jit
-    def _loss(fac, target):
-        return loss(fac(), target)
-    gradient = ab.jit(ab.grad(_loss))
+    loss_and_grad = ab.loss_and_grad(loss, opt_fac, target)
 
+    # bookkeeping
+    best_factors = [np.zeros_like(ab.to_numpy(x)) for x in opt_fac.factors]
     best_loss = np.ones(nvec) * np.inf
     converged = np.zeros(nvec)
     pbar = tqdm.tqdm(total=max_steps + 1)
 
     for step in range(max_steps):
         pbar.update(1)
-        opt.step(gradient(opt_fac, target).factors)
-
-        if step % round(max_steps/20) == 0:
-            # update best factorization
-            curr_loss = loss(opt_fac(), target, sum_vec=False)
-            new_best = []
-            for b, o in zip(best_fac.factors, opt_fac.factors):
-                for i, l in enumerate(zip(curr_loss, best_loss)):
-                    if l[0] < l[1]:
-                        if b.shape[-1] == 1:
-                            # TODO: this weird way of updating is required
-                            # by JAX
-                            b = b.at[..., 0].set(o[..., 0])
-                        else:
-                            b = b.at[..., i].set(o[..., i])
-                        if l[0] < tol:
-                            converged[i] = 1
-                new_best.append(b)
-            best_fac.factors = new_best
-
-            if stop_by == 'first':
-                if np.any(converged):
-                    pbar.update(max_steps - step)
-                    break
-            elif isinstance(stop_by, int):
-                if np.count_nonzero(converged) >= stop_by:
-                    pbar.update(max_steps - step)
-                    break
-            else:
-                if stop_by is not None:
-                    raise RuntimeError(
-                        f'Invalid argument value for stop_by: {stop_by}'
-                    )
+        _, grad = loss_and_grad(opt_fac, target)
+        with ab.no_grad():
+            opt.step(grad)
+
+            if step % round(max_steps/20) == 0:
```

Comment:
It might feel a bit arbitrary if we are updating the best solution by every `round(max_steps/20)`. How about making it an argument e.g. `checkpoint_freq` that defaults to 1?

Reply:
I have added this argument but set the default value to 50. Checking every single iteration by default seemed a bit much to me.